### PR TITLE
Match with state - fixes #121

### LINKED ISSUE #121: Allow matching with static functions
One issue with `.Match` is that it allocated objects for case matching functions to capture the closure. C# has a mechanism to prevent this from happening with a `static` keyword on lambda. However, for it to work the "closure" state has to be threaded manually.

Consider an example of such a match function called `.FoldMatch` based on your Expression example implemented in the `Expression` class. Functions `StaticEvaluate` and `EvaluateUniqueVar` show how such a utility could be used:
```csharp
// Adapted from https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/discriminated-unions#using-discriminated-unions-for-tree-data-structures

using System.Collections.Immutable;
using Dunet;
using static Expression;

var environment = new Dictionary<string, int>()
{
    ["a"] = 1,
    ["b"] = 2,
    ["c"] = 3,
};

var expression = new Add(new Variable("a"), new Multiply(new Number(2), new Variable("b")));

// Evaluate a + 2 * b
var result = Evaluate(environment, expression);

Console.WriteLine(result); // 5

static int Evaluate(Dictionary<string, int> env, Expression exp) =>
    exp.Match(
        number => number.Value,
        add => Evaluate(env, add.Left) + Evaluate(env, add.Right),
        multiply => Evaluate(env, multiply.Left) * Evaluate(env, multiply.Right),
        variable => env[variable.Value]
    );

// Allows to evaluate expressions without allocating lambdas
static int StaticEvaluate(Dictionary<string, int> env, Expression exp) =>
    exp.FoldMatch(
        env,
        static (_, number) => number.Value,
        static (e, add) => StaticEvaluate(e, add.Left) + StaticEvaluate(e, add.Right),
        static (e, multiply) => StaticEvaluate(e, multiply.Left) * StaticEvaluate(e, multiply.Right),
        static (e, variable) => e[variable.Value]
    );

// An example of how to thread state through a recursive function
// It simulates a (made up) use case where variables can be used only once
static (IImmutableDictionary<string, int> state, int result) 
    EvaluateUniqueVar(IImmutableDictionary<string, int> env, Expression exp) =>
    exp.FoldMatch(
        env,
        static (e, number) => (e, number.Value),
        static (e, add) =>
        {
            (e, var left) = EvaluateUniqueVar(e, add.Left);
            (e, var right) = EvaluateUniqueVar(e, add.Right);
            
            return (e, left + right);
        },
        static (e, multiply) =>
        {
            (e, var left) = EvaluateUniqueVar(e, multiply.Left);
            (e, var right) = EvaluateUniqueVar(e, multiply.Right);
            
            return (e, left * right);
        },
        static (e, variable) => (e.Remove(variable.Value), e[variable.Value])
    );

[Union]
public partial record Expression
{
    partial record Number(int Value);

    partial record Add(Expression Left, Expression Right);

    partial record Multiply(Expression Left, Expression Right);

    partial record Variable(string Value);
    
    public T FoldMatch<T, TState>(
        TState state,
        Func<TState, Number, T> number,
        Func<TState, Add, T> add,
        Func<TState, Multiply, T> multiply,
        Func<TState, Variable, T> variable
    ) => this switch
    {
        Number x   => number(state, x),
        Add x      => add(state, x),
        Multiply x => multiply(state, x),
        Variable x => variable(state, x),
        _          => throw new InvalidOperationException()
    };
}
```

### PR DESCRIPTION
**Pull request by panoukos41**
Hey @domn1995 I have created an implementation that should provide a solution for #121 that generates `Variant` & `Specific` `Match` methods that accept custom state eg:

```csharp
// Variant
public abstract TMatchOutput Match<TState, TMatchOutput>(TState state, ...);
public abstract void `Match<TState>(TState state, ...);

// Specific
public abstract TMatchOutput MatchSpecific<TState, TMatchOutput>(TState state ...);
public abstract void MatchSpecific<TState>(TState state ...);
```

I am also creating this as a draft since I would like some guidance about the tests and implementation.

For the implementation: I implemented the state methods together with the non-state ones but I could also move them to a different method. I just thought they belong together with their non-state counterparts.

For the tests: I know I could change some tests that execute methods to also execute the state counterpart or I could create a new test class that does just that but I am not sure which way to go 😄

### CODE REVIEW

**Comment by domn1995** [comment]
> For the implementation: I implemented the state methods together with the non-state ones but I could also move them to a different method. I just thought they belong together with their non-state counterparts.

I think this works well right now. The source builder is getting pretty long, but we can clean that up as a separate pull request later 👍

> For the tests: I know I could change some tests that execute methods to also execute the state counterpart or I could create a new test class that does just that but I am not sure which way to go 

My gut says to split it up into its own "state matching" test class. This way, we don't couple any of the existing test suites to the new one. Let me know if you need any help with those tests. :)

**Comment by panoukos41** [comment]
Thanks for your fast response and sorry for not getting back quickly It was midnight here in Greece 🇬🇷.

I will give it a try later today and if I have trouble finding which tests to mimic I will let you know 😁

**Comment by panoukos41** [comment]
@domn1995 I pushed the tests let me know if something needs changing or how I could help more 😄 otherwise :shipit:! (always wanted to use this emoji :D)

File: src/UnionGeneration/UnionSourceBuilder.cs [hunk 1]
```diff
@@ -1,4 +1,5 @@
-﻿using System.Text;
+﻿using System.Net.NetworkInformation;
```
Comment (domn1995) [inline src/UnionGeneration/UnionSourceBuilder.cs:1]:
```suggestion
```
Looks like a rogue import

**Review by domn1995 — CHANGES_REQUESTED** [review]

**Review by domn1995 — APPROVED** [review]

**Comment by domn1995** [comment]
Everything look awesome! Thanks again for the contribution! :shipit: :shipit: :shipit:

**Pull request merged**